File: cpp/2923_find_champion/find_champion.cpp
/*
 * @Date: 2024-04-12
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-04-12
 * @FilePath: /algorithm/cpp/2923_find_champion/find_champion.cpp
 */

#include <cassert>
#include <numeric>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int findChampion(vector<vector<int>>& grid) {
    int n = grid.size();
    for (int i = 0; i < n; i++) {
      if (accumulate(grid[i].begin(), grid[i].end(), 0) == n - 1) {
        return i;
      }
    }
    return -1;
  }
};

int main() {
  vector<tuple<vector<vector<int>>, int>> tests{
      {{{0, 1}, {0, 0}}, 0},
      {{{0, 0, 1}, {1, 0, 1}, {0, 0, 0}}, 1},
  };

  for (auto& [grid, ans] : tests) {
    assert(Solution().findChampion(grid) == ans);
  }
}

File: rust/2924_find_champion/find_champion.rs
/*
 * @Date: 2024-04-13
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-04-13
 * @FilePath: /algorithm/rust/2924_find_champion/find_champion.rs
 */

struct Solution;

impl Solution {
    pub fn find_champion(n: i32, edges: Vec<Vec<i32>>) -> i32 {
        let mut degree = vec![0; n as usize];
        for e in edges.iter() {
            degree[e[1] as usize] += 1;
        }
        let mut champion = -1;
        for i in 0..n {
            if degree[i as usize] == 0 {
                if champion == -1 {
                    champion = i;
                } else {
                    return -1;
                }
            }
        }
        champion
    }
}

fn main() {
    let tests = vec![
        (3, vec![vec![0, 1], vec![1, 2]], 0),
        (4, vec![vec![0, 2], vec![1, 3], vec![1, 2]], -1),
    ];

    for (n, edges, ans) in tests {
        assert_eq!(Solution::find_champion(n, edges), ans);
    }
}
